### PR TITLE
Update email when linking respondent

### LINKED ISSUE #2269: Add missing email address when linking survey response
## Description
If an email address is missing in the person object but provided by the survey survey response, the user should be able to update its email address when linking the response.

## Estimated size (S, M, L)

I think this issue is S because it's a simple operation.

## Possible implementations
- Add a modal asking user to add the email to the person, right after the person has been selected for linking.

## Design specifications

image

### PR DESCRIPTION
**Pull request by rebecarubio**
## Description
This PR adds a dialog to allow the user to add the respondent email to the person that is going to be link as a new respondent. 
It also introduces a new extra reducer in the `Survey ` store that updates de submission list when a person is updated. 

## Screenshots

https://github.com/user-attachments/assets/f8299132-391a-4a55-84f7-46f0d25cd920

## Changes
* Adds an `if` statement in `SurveySubmissionsList` when a person to be linked as a respondent doesn't have an email
* Adds a `SurveyDialogLink `component to display a Dialog when the if statement is triggered 
* Adds an extra reducer in `Survey `store to update the submissions list when `personUpdated` action is triggered. 

## Notes to reviewer
None

## Related issues
Resolves #2269

### CODE REVIEW

File: src/features/surveys/components/SurveySubmissionsList.tsx [hunk 1]
```diff
@@ -198,6 +208,12 @@ const SurveySubmissionsList = ({
         id: row.id,
       });
       setRespondentId(person?.id || null);
+
+      const personHasNoEmail = person?.id !== null && person?.email === null;
+      const respondentEmail = row.respondent?.email;
+      if (personHasNoEmail && respondentEmail !== undefined) {
```
Comment (richardolsson) [inline src/features/surveys/components/SurveySubmissionsList.tsx:214]: Whenever I see strict equality checks (`===` and `!==`) in typescript, that is a bit of a yellow flag for me.

### Why strict equality?
The purpose of the strict equality operators in javascript (as opposed to  the regular equality operators, `==` and `!=`) is to compare the values without making any type conversions. The regular operator is "smart" in that it tries to convert the operands to types that can be compared, but in doing so it sometimes does the wrong thing. For example, `null == undefined` is true (because one or both operands are converted to a compatible type), but `null === undefined` is not (because they are different types). Similarly, `1 == '1'` is true, because the left operand is converted to a string before comparing, but `1 === '1'` is not true, because the left operand is a number and the right is a string.

The strict equality is important in javascript, because types are so dynamic that you're encouraged not to care about what type something will be at execution time. Using strict equality is a good default in javascript, because it lets you know (by not working as intended) if something is of a different type than you expected. 

For example, imagine this JS code:

```js
if (a == 5) {
  const eight = a + 3;
  console.log(eight);
}
```

Will it print 8? We can't know! Because if `a` is `'5'` (the string), the boolean expression will evaluate to true (because `'5' == 5`) but `eight` will actually be `'53'` (the string resulting from concatenating the two strings '5' and '3').

In this case, the "cleverness" of the default equality operator in JS is biting us. If we had used strict equality (`===`) the if statement would not have executed.

That is probably _also not what we want_, but at least it's "jarring" enough for the tester/user to hopefully discover that there is some problem here, and the programmer can go and find the original flaw, i.e. whatever is causing `a` to be a string when the code assumes it's a number.

That's why using `===` in javascript is a good idea. It helps you find bugs related to types, but it doesn't fix them.

### Strict equality and Typescript
Typescript, however, is different. Although typescript is of course only javascript when it gets executed, the problem of not knowing types is (mostly) avoided by compile-time type-checking.

Typescript type-checks operands in equality expressions. This means that if the two operands have different types, typescript will fail to compile. You can [see this in action here](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAhjAvDA5AVhQbgFAEsBmMAFAosmgJQwDe2MMoksAprgOYAWsyCA1DAGYc9RhBAAbZgDpxINkVacoFHAF9s2UbABGSGGhwFiusvqq0R4aDEVc9u-kLoMrE6bPm3laoA):

image

The bugs that strict equality helps us find _at runtime_ (if we're lucky), typescript helps us find _at compile-time_ (always!).

This does not mean that strict equality never has a place in typescript. There are cases where typescript will allow a comparison between two separate (but related) types. For example, the expression `null == undefined` is valid in typescript and once executed in javascript will evaluate to true. So if you really want to distinguish `null` from `undefined`, you need strict equality.

Which leads us to your code.

### The problem(?) with your code
Your code is comparing `respondentEmail` with `undefined` using strict equality. This means that the if statement will only execute if `respondentEmail` is anything other than strictly `undefined`.

If `respondentEmail` is `null`, I would argue (in this context) that it's conceptually equal to `undefined`. The purpose of the code is to express some intent. In this case, the intent is to show a dialog if the respondent _has no email_, which at least in theory could mean that it's either `undefined` or `null`.

Now, I don't think that in this specific case `respondentEmail` can ever be `null`, because `resondent.email` will always be a string if it exists, so there is no bug here.

But if we ever make changes in the future to the type of `ZetkinSurveySubmission`, to allow `respondent.email` to be `null | string`, a bug would be introduced in this completely separate part of the code. Because even if we change the type, the _intent_ is still to open this dialog if the respondent _has no email_, but that won't happen because the email is not strictly `undefined`, which is what you're checking for here.

So, in conclusion, this code is being unnecessarily strict, probably due to an instinct conditioned from years of javascript programming, but you're not getting any of the benefits from that instinct, only the (potential) problems.

**Review by richardolsson — CHANGES_REQUESTED** [review]
Nice work! I've seen this work on your screen and I thought it looked great. But then at first I couldn't get it to work for me. Here's what I tried:

1. With the preview build, I went to ["Simon Zetterlund"](https://app-zetkin-joh4ktbwb-zetkin.vercel.app/organize/1/people/40) and removed his email. Note that this sets the `email` to an empty string, which is probably the cause of this bug.
2. I then went to [this survey submissions list](https://app-zetkin-joh4ktbwb-zetkin.vercel.app/organize/1/projects/standalone/surveys/1/submissions) where there is a submission made using my email address.
3. I tried linking the submission to Simon Zetterlund, which works, but the modal does not open.

I then tried creating a new person instead of using an existing one, and not setting an email to begin with, which does work. It probably works because the email will be `null`.

I also think that it would be nice to get a designer to look at the modal. Maybe you can just pair up with someone in the office?

Finally, I left a bunch of comments below. The code looks mostly great, but there are some minor things that I think can be improved, and also some that I identified as a learning opportunity – hopefully appreciated. 😊

File: src/features/surveys/components/SurveyDialogLink.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+} from '@mui/material';
+
+import messageIds from '../l10n/messageIds';
+import { useNumericRouteParams } from 'core/hooks';
+import { useMessages } from 'core/i18n';
+import usePersonMutations from 'features/profile/hooks/usePersonMutations';
+import { ZetkinPerson } from 'utils/types/zetkin';
+import ZUIPersonAvatar from 'zui/ZUIPersonAvatar';
+
+const SurveyDialogLink = ({
```
[the remark this answers is not in the document]
Comment (rebecarubio) [reply src/features/surveys/components/SurveyDialogLink.tsx:19]: Agree, maybe SurveyLinkDialog or SurveyRespondentDialog?

[the remark this answers is not in the document]
Comment (richardolsson) [reply src/features/surveys/components/SurveyDialogLink.tsx:19]: I guess `SurveyLinkDialog` works. Good suggestion!

File: src/features/surveys/components/SurveySubmissionsList.tsx (hunk 1 above)
Comment (rebecarubio) [reply src/features/surveys/components/SurveySubmissionsList.tsx:214]: Thank you so much for this detailed explanation and thoughtful feedback! I genuinely appreciate the time you took to not only review the code but also provide a deep dive into the reasoning behind strict equality checks in both JavaScript and TypeScript.
It’s clear that my instinct for writing `===` here came from JavaScript habits, but, as you said, the TypeScript compiler already ensures type safety. I’ll update the code and I will keep this lesson in mind for the future. Your mentorship is so great! 🙌

File: src/features/surveys/components/SurveySubmissionsList.tsx [hunk 3]
```diff
@@ -30,6 +31,15 @@ const SurveySubmissionsList = ({
...
   const { orgId } = useRouter().query;
   const { openPane } = usePanes();
+  const [dialogOpen, setDialogOpen] = useState(false);
+  const [dialogPerson, setDialogPerson] = useState<ZetkinPerson>();
+  const [dialogEmail, setDialogEmail] = useState('');
+
+  const handleUpdateEmail = (option: ZetkinPerson, email: string) => {
+    setDialogEmail(email);
+    setDialogPerson(option);
+    setDialogOpen(true);
```
[the remark this answers is not in the document]
Comment (rebecarubio) [reply src/features/surveys/components/SurveySubmissionsList.tsx:41]: I've removed the setDialogOpen, so that's not needed! But I do need the other ones :)

File: src/features/surveys/components/SurveySubmissionsList.tsx [hunk 4]
```diff
@@ -209,10 +203,13 @@ const SurveySubmissionsList = ({
       });
       setRespondentId(person?.id || null);
 
-      const personHasNoEmail = person?.id !== null && person?.email === null;
       const respondentEmail = row.respondent?.email;
-      if (personHasNoEmail && respondentEmail !== undefined) {
-        handleUpdateEmail(person, respondentEmail);
+      if (person) {
+        const personHasNoEmail = person.email == null || person.email == '';
+        if (personHasNoEmail && respondentEmail != undefined) {
```
Comment (richardolsson) [inline src/features/surveys/components/SurveySubmissionsList.tsx:209]: I think this is much nicer! I believe it could have been made even simpler, by relying on the fact that `null` and `undefined` are falsy, like so:

```suggestion
      if (person) {
        const personHasNoEmail = !person.email || person.email == '';
        if (personHasNoEmail && respondentEmail) {
```

But that's not worth changing the code over!

**Review by richardolsson — APPROVED** [review]
Nice work! Let's get this merged. 💯

File: src/features/surveys/components/SurveySubmissionsList.tsx (hunk 1 above)
Comment (xela1601) [reply src/features/surveys/components/SurveySubmissionsList.tsx]: @richardolsson thanks from my side as well for this detailed explanation that i just read through.
You could make something bigger out of this (like a blog post or smth like that).
It's a pity for this to be forgotten in this PR comment 😅 

I was just wondering why using a comparison like `respondentEmail !== undefined` if you could compare it with the whole value `respondentEmail` or to be more specific the `truthyness` (?) of this value instead?

Comment (xela1601) [reply src/features/surveys/components/SurveySubmissionsList.tsx]: Edit: after scrolling down a bit, this is exactly what you suggested as you approved this PR 😅
